File: src/main/java/Searcher.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;

public class Searcher {

    private InvertedIndex invertedIndex;
    private SearchingExpression searchingExpression;

    public void run(String folderAddress) {
        invertedIndex = new InvertedIndex();
        invertedIndex.indexAllFiles(folderAddress);
        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.println("enter a word for search or EXIT to exit:");
            String input = scanner.nextLine();
            if (input.equals("EXIT")) return;
            searchingExpression = new SearchingExpression(input);
            printResults(search());
            System.out.println("---------------------------------------------------");
        }
    }

    public List<WordInfo> search() {
        List<WordInfo> candidates;

        int navigatingIndex = 0;
        try {
            while (InvertedIndex.getStopWords().contains(searchingExpression.getWords().get(navigatingIndex)))
                navigatingIndex++;
            candidates = new LinkedList<>(searchForAWord(searchingExpression.getWords().get(navigatingIndex)));
        } catch (Exception e) {
            return null;
        }
        int ignoredCounter = 0;
        for (navigatingIndex += 1; navigatingIndex < searchingExpression.getWords().size(); navigatingIndex++) {
            String word = searchingExpression.getWords().get(navigatingIndex);
            if (InvertedIndex.getStopWords().contains(word)) {
                ignoredCounter++;
                continue;
            }
            List<WordInfo> demo = searchForAWord(searchingExpression.getWords().get(navigatingIndex));
            reduceResultsToMatchSearch(candidates, ignoredCounter, demo);
            ignoredCounter = 0;
        }
        candidatesProcessor(candidates);
        return candidates;
    }

    private void candidatesProcessor(List<WordInfo> candidates) {
        HashMap<String, WordInfo> allCandidates = new HashMap<>();
        handlePlusWords(allCandidates);
        sumResultsWithPlusWords(allCandidates, candidates);
        candidates = new LinkedList<>(allCandidates.values());
        deleteMinusWordsFromCandidates(candidates);
    }

    private void reduceResultsToMatchSearch(List<WordInfo> candidates, int ignoredCounter, List<WordInfo> demo) {
        for (int j = candidates.size() - 1; j >= 0; j--) {
            WordInfo candidate = candidates.get(j);
            boolean isExist = false;

            for (WordInfo wordInfo : demo) {
                if (wordInfo.getFileName().equals(candidate.getFileName())
                        && candidate.getPosition() + ignoredCounter + 1 == wordInfo.getPosition()) {
                    candidates.set(j, wordInfo);
                    isExist = true;
                    break;
                }
            }
            if (!isExist)
                candidates.remove(j);
        }
    }

    private void sumResultsWithPlusWords(HashMap<String, WordInfo> allCandidates, List<WordInfo> candidates) {
        for (WordInfo candidate : candidates) {
            allCandidates.put(candidate.getFileName(), candidate);
        }
    }

    private void handlePlusWords(HashMap<String, WordInfo> allCandidates) {
        for (String plusWord : searchingExpression.getPlusWords()) {
            for (WordInfo wordInfo : searchForAWord(plusWord)) {
                allCandidates.put(wordInfo.getFileName(), wordInfo);
            }
        }
    }

    private void printResults(List<WordInfo> candidates) {
        if (candidates == null || candidates.size() == 0) {
            System.out.println(ConsoleColors.ANSI_RED + "please try a different keyword for your search!"
                    + ConsoleColors.ANSI_RESET);
            return;
        }
        for (WordInfo candidate : candidates)
            System.out.println("File name: " + ConsoleColors.ANSI_CYAN + candidate.getFileName() + ConsoleColors.ANSI_RESET
                    + " ApproximatePosition: " + ConsoleColors.ANSI_GREEN + (candidate.getPosition() - candidates.size() + 1)
                    + ConsoleColors.ANSI_RESET);
    }


    private List<WordInfo> searchForAWord(String word) {
        word = word.toLowerCase();
        return invertedIndex.getIndex().get(word);
    }

    private void deleteMinusWordsFromCandidates(List<WordInfo> candidates) {
        for (String minusWord : searchingExpression.getMinusWords()) {
            List<WordInfo> toBeRemovedDocs = searchForAWord(minusWord);
            for (WordInfo toBeRemovedDoc : toBeRemovedDocs)
                for (int j = candidates.size() - 1; j >= 0; j--)
                    if (candidates.get(j).getFileName().equals(toBeRemovedDoc.getFileName()))
                        candidates.remove(j);
        }
    }

}

File: Phase11_1/SampleLibrary/SampleLibrary/Searcher.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SampleLibrary
{
    public class Searcher
    {
        public Searcher(SearchContext searchContext, InvertedIndex invertedIndex)
        {
            SearchContext = searchContext;
            InvertedIndex = invertedIndex;
        }

        public InvertedIndex InvertedIndex { get; }
        private SearchContext SearchContext { get; }

        public List<WordInfo> Search(string searchingExpression)
        {
            var searchingExpressionLowered = searchingExpression.ToLower();
            var allCandidates =
                InstantiateRequiredLists(searchingExpressionLowered, out var words,
                    out var plusWords, out var minusWords);

            OperatorWordsHandler.IsolatePlusAndMinusWords(words, plusWords, minusWords);

            if (FindFirstNonStopWord(words, out var navigatingIndex))
            {
                return new List<WordInfo>();
            }

            var candidates = FindCandidates(words, navigatingIndex);
            if (candidates.Count == 0)
            {
                return candidates;
            }

            var ignoredCounter = 0;
            for (navigatingIndex += 1; navigatingIndex < words.Count; navigatingIndex++)
            {
                var word = words[navigatingIndex];
                if (InvertedIndex.StopWords.Contains(word))
                {
                    ignoredCounter++;
                    continue;
                }

                var demo = SearchForAWord(words[navigatingIndex]);
                ReduceResultsToMatchSearch(candidates, ignoredCounter, demo);
                ignoredCounter = 0;
            }

            OperatorWordsHandler.HandlePlusWords(allCandidates, plusWords, this);
            return OperatorWordsHandler.HandlePlusAndMinusWords(allCandidates, candidates, minusWords, this);
        }

        private List<WordInfo> FindCandidates(IReadOnlyList<string> words, int navigatingIndex)
        {
            ICollection<WordInfo> candidates = SearchForAWord(words[navigatingIndex]);

            return new List<WordInfo>(candidates);
        }

        private static Dictionary<string, WordInfo> InstantiateRequiredLists(string searchingExpression,
            out List<string> words, out List<string> plusWords,
            out List<string> minusWords)
        {
            var allCandidates = new Dictionary<string, WordInfo>();
            words = new List<string>(searchingExpression.Split(' '));
            plusWords = new List<string>();
            minusWords = new List<string>();
            return allCandidates;
        }

        private bool FindFirstNonStopWord(List<string> words, out int navigatingIndex)
        {
            navigatingIndex = 0;
            try
            {
                while (InvertedIndex.StopWords.Contains(words[navigatingIndex])) navigatingIndex++;
            }
            catch (Exception)
            {
                Console.WriteLine("please try a different keyword for your search!");
                return true;
            }

            return false;
        }

        private void ReduceResultsToMatchSearch(IList<WordInfo> candidates, int ignoredCounter,
            IReadOnlyCollection<WordInfo> demo)
        {
            for (var j = candidates.Count - 1; j >= 0; j--)
            {
                var candidate = candidates[j];
                var isExist = false;

                foreach (var wordInfo in demo.Where(wordInfo => wordInfo.FileName == candidate.FileName
                                                                && candidate.Position + ignoredCounter + 1 ==
                                                                wordInfo.Position))
                {
                    candidates[j] = wordInfo;
                    isExist = true;
                    break;
                }

                if (!isExist)
                {
                    candidates.RemoveAt(j);
                }
            }
        }

        public void PrintResults(List<WordInfo> candidates)
        {
            if (candidates == null)
            {
                Console.WriteLine("there is no match!");
                return;
            }

            candidates.ForEach(candidate =>
                Console.WriteLine("File name: " + candidate.FileName + " ApproximatePosition: " +
                                  +(candidate.Position - candidates.Count + 1)));
        }

        public List<WordInfo> SearchForAWord(string word)
        {
            word = word.ToLower();
            var wordIds = SearchContext.Words.Where(x => x.WordContent == word).Select(x => x.WordId);
            return SearchContext.WordInfos.Where(y => wordIds.Contains(y.WordId)).ToList();
        }

    }
}